src/ArraySetops.chpl

```diff
@@ -114,30 +122,149 @@ module ArraySetops
+            const bDom = b.localSubdomain();
+
+            // the ternaries allow me to declare the next variables as const
+            // info for pushing to previous locale
+            const numSendBack = if loc == Locales[0] then 0 else localizedPullLocalCount[here.id - 1];
+            const sendBack = if loc == Locales[0] then (false, false) else localizedPullLocalFlag[here.id - 1];
+            // info for fetching from next locale
+            const numPullLocal = localizedPullLocalCount[here.id];
+            const pullLocal = pullLocalFlag[here.id];
+
+            // skip any indices that were pulled to previous locale
+            const aLow = if sendBack[0] then (aDom.first + numSendBack) else aDom.first;
+            const bLow = if sendBack[1] then (bDom.first + numSendBack) else bDom.first;
+            const aSlice = aLow..<(aDom.last + 1);
+            const bSlice = bLow..<(bDom.last + 1);
+
+            var tmp: [0..<(aDom.size + bDom.size - numSendBack + numPullLocal)] (t, t2);
+
+            // write the local part of a (minus any skipped values) into tmp
+            tmp[0..<aSlice.size] = [(key,val) in zip(a.localSlice[aSlice], aVal.localSlice[aSlice])] (key,val);
+            // write the local part of b (minus any skipped values) into tmp
+            tmp[aSlice.size..#bSlice.size] = [(key,val) in zip(b.localSlice[bSlice], bVal.localSlice[bSlice])] (key,val);
+
+            // add any indices that we need to pull to this locale
+            // this not a local slice, but it should (hopefully) take advantage of bulk transfer
+            if pullLocal[0] {
+              const pullFromASlice = (aDom.last+1)..#numPullLocal;
+              tmp[(aSlice.size+bSlice.size)..#(numPullLocal+1)] = [(key,val) in zip(a[pullFromASlice], aVal[pullFromASlice])] (key,val);
+            }
+            else if pullLocal[1] {
+              const pullFromBSlice = (bDom.last+1)..#numPullLocal;
+              tmp[(aSlice.size+bSlice.size)..#(numPullLocal+1)] = [(key,val) in zip(b[pullFromBSlice], bVal[pullFromBSlice])] (key,val);
+            }
+
+            // run chpl's builtin parallel radix sort for local arrays with a comparator defined in
+            // RadixSortLSD.chpl which sorts on the key
+            twoArrayRadixSort(tmp, new KeysRanksComparator());
+
+            const writeToResultSlice = (segs[here.id]+numSendBack)..#tmp.size;
+            sortedIdx[writeToResultSlice] = [(key, val) in tmp] key;
+            permutedVals[writeToResultSlice] = [(key, val) in tmp] val;
+          }
+        }
+      }
+      else {
+        var perm = makeDistArray(allocSize, int);
+        forall (s, p, sp) in zip(sortedIdx, perm, radixSortLSD(concatArrays(idx1, idx2, ordered=false))) {
+          (s, p) = sp;
+        }
+        const vals = concatArrays(val1, val2, ordered=false);
+        forall (p, i) in zip(permutedVals, perm) with (var agg = newSrcAggregator(t2)) {
+          agg.copy(p, vals[i]);
```

Comment:
excellent question!

So for:
```chapel 
forall (s, p, sp) in zip(sortedIdx, perm, radixSortLSD(concatArrays(idx1, idx2, ordered=false))) {
  (s, p) = sp;
}
```
Firstly we know arrays that are distributed the same way because they are the same size and block distibuted (If they weren't the same size, chpl would yell at us about trying to zip them together). So this means that `s`, `p`, and `sp` will always be on the same locale as each other (i.e. they are all local to each other). Since they're on the same locale, there's no comm and no need for aggregation. Creating aggregators carries a bit overhead, so it's best to avoid if it's not necessary.

Note: this is a big reason why I'm a fan of using looping variables instead of indexing into the arrays since it makes it easier to tell at a glance that these values are local to the locale where the computation is happening. (also i just think it looks cleaner). We could rewrite this as the below (and it would be equivalent), but it would be tougher to know what was local:
```chapel
var sortedArr = radixSortLSD(concatArrays(idx1, idx2, ordered=false));
forall i in sortedIdx.domain {
  (sortedIdx[i], perm[i]) = sortedArr[i];
}
```

for the example in the comment:
```chapel
forall (p, i) in zip(permutedVals, perm) with (var agg = newSrcAggregator(t2)) {
  agg.copy(p, vals[i]);
}
```
Using the reasoning in above case, we are confident that `p` and `i` are local to one another. But we don't know that `p` and `vals[i]` live on the same locale. if `vals[i]` lives on a different locale (and there's a very good chance it does) then we want to avoid the comm of writing `p` int `vals[i]`. So using an aggregator allows us to wait until we have a couple that need to go to a locale and send them all over at once

Onces again this highlights how nice following the convention of looping variables / zipping is when possible because the fact that we are indexing into vals is an indicator that it could be remote (i.e. living on a different local)

Reply:
Great explanation.  Thanks!